include/deal.II/matrix_free/mapping_info.templates.h

```diff
@@ -370,6 +370,52 @@ namespace internal
 
 
 
+    template <int dim, typename Number, typename VectorizedArrayType>
+    void
+    MappingInfo<dim, Number, VectorizedArrayType>::recategorize(
+      std::vector<std::pair<unsigned int, unsigned int>> cell_level_index,
+      const FaceInfo<VectorizedArrayType::size()> &      faces)
+    {
+      constexpr unsigned int n_lanes = VectorizedArrayType::size();
+      const unsigned int     n_quads = face_data_by_cells.size();
+
+      // Resize types
+      cell_type.resize(cell_level_index.size() / n_lanes);
+      face_type.resize(faces.faces.size());
+
+      // loop over all quad formulas
+      for (uint my_quad = 0; my_quad < n_quads; ++my_quad)
+        {
+          // Set data index offset to zero, quad to iterator value
+          cell_data[my_quad].data_index_offsets.resize(cell_type.size());
+          cell_data[my_quad].quadrature_point_offsets.resize(cell_type.size());
+
+          for (uint i = 0; i < cell_data[my_quad].data_index_offsets.size();
+               ++i)
+            {
+              cell_data[my_quad].data_index_offsets[i]       = 0;
+              cell_data[my_quad].quadrature_point_offsets[i] = i;
+            }
+
+          // Set data index offset to zero, quad to iterator value
+          face_data[my_quad].data_index_offsets.resize(face_type.size());
+          face_data_by_cells[my_quad].data_index_offsets.resize(
+            cell_type.size() * GeometryInfo<dim>::faces_per_cell);
+          face_data_by_cells[my_quad].quadrature_point_offsets.resize(
+            cell_type.size() * GeometryInfo<dim>::faces_per_cell);
+
+          for (uint i = 0; i < face_data[my_quad].data_index_offsets.size();
+               ++i)
+            {
+              face_data[my_quad].data_index_offsets[i]                = 0;
+              face_data_by_cells[my_quad].data_index_offsets[i]       = i;
+              face_data_by_cells[my_quad].quadrature_point_offsets[i] = i;
```

Comment:
Indeed, this is a problem...

Reply:
Out of curiosity, what would have been the speedup?

Reply:
I currently spend 2/3 of my overall simulation time with reinitialization in release mode i.e. it would be three times faster:
```
+---------------------------------------+------------------+------------+------------------+
| Total wallclock time elapsed          |      1089s     0 |      1089s |      1089s     0 |
|                                       |                  |                               |
| Section                   | no. calls |   min time  rank |   avg time |   max time  rank |
+---------------------------------------+------------------+------------+------------------+
| Flag cells for FV         |     12369 |    0.2602s     0 |    0.2602s |    0.2602s     0 |
| compute smoothness        |     12369 |     23.54s     0 |     23.54s |     23.54s     0 |
| compute transport speed   |      2063 |     4.892s     0 |     4.892s |     4.892s     0 |
| output                    |        26 |     35.93s     0 |     35.93s |     35.93s     0 |
| recategorize/reinitialize |     12368 |     698.3s     0 |     698.3s |     698.3s     0 |
| rk_stage cell centric     |     61840 |     324.6s     0 |     324.6s |     324.6s     0 |
+---------------------------------------+------------------+------------+------------------+
```